File: be/src/io/fs/file_system.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "io/fs/file_system.h"

#include <bthread/bthread.h>

#include "io/fs/file_reader.h"

namespace doris {
namespace io {

Status FileSystem::create_file(const Path& file, FileWriterPtr* writer,
                               const FileWriterOptions* opts) {
    DCHECK(bthread_self() == 0) << "FileSystem::create_file should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(file, path));
    return create_file_impl(path, writer, opts);
}

Status FileSystem::open_file(const Path& file, FileReaderSPtr* reader,
                             const FileReaderOptions* opts) {
    DCHECK(bthread_self() == 0) << "FileSystem::open_file should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(file, path));
    return open_file_impl(path, reader, opts);
}

Status FileSystem::create_directory(const Path& dir, bool failed_if_exists) {
    DCHECK(bthread_self() == 0) << "FileSystem::create_directory should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(dir, path));
    return create_directory_impl(path, failed_if_exists);
}

Status FileSystem::delete_file(const Path& file) {
    DCHECK(bthread_self() == 0) << "FileSystem::delete_file should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(file, path));
    return delete_file_impl(path);
}

Status FileSystem::delete_directory(const Path& dir) {
    DCHECK(bthread_self() == 0) << "FileSystem::delete_directory should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(dir, path));
    return delete_directory_impl(path);
}

Status FileSystem::batch_delete(const std::vector<Path>& files) {
    DCHECK(bthread_self() == 0) << "FileSystem::batch_delete should not be called in bthread";
    std::vector<Path> abs_files;
    for (const auto& file : files) {
        Path abs_file;
        RETURN_IF_ERROR(absolute_path(file, abs_file));
        abs_files.push_back(abs_file);
    }
    return batch_delete_impl(abs_files);
}

Status FileSystem::exists(const Path& path, bool* res) const {
    DCHECK(bthread_self() == 0) << "FileSystem::exists should not be called in bthread";
    Path fs_path;
    RETURN_IF_ERROR(absolute_path(path, fs_path));
    return exists_impl(fs_path, res);
}

Status FileSystem::file_size(const Path& file, int64_t* file_size) const {
    DCHECK(bthread_self() == 0) << "FileSystem::file_size should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(file, path));
    return file_size_impl(path, file_size);
}

Status FileSystem::list(const Path& dir, bool only_file, std::vector<FileInfo>* files,
                        bool* exists) {
    DCHECK(bthread_self() == 0) << "FileSystem::list should not be called in bthread";
    Path path;
    RETURN_IF_ERROR(absolute_path(dir, path));
    return list_impl(path, only_file, files, exists);
}

Status FileSystem::rename(const Path& orig_name, const Path& new_name) {
    DCHECK(bthread_self() == 0) << "FileSystem::rename should not be called in bthread";
    Path orig_path;
    RETURN_IF_ERROR(absolute_path(orig_name, orig_path));
    Path new_path;
    RETURN_IF_ERROR(absolute_path(new_name, new_path));
    return rename_impl(orig_path, new_path);
}

} // namespace io
} // namespace doris

File: fe/fe-core/src/main/java/org/apache/doris/fs/FileSystem.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.fs;

import org.apache.doris.fs.io.DorisInputFile;
import org.apache.doris.fs.io.DorisOutputFile;

import java.io.Closeable;
import java.io.IOException;
import java.util.Collection;
import java.util.Set;

/**
 * Core filesystem abstraction for Apache Doris.
 * <p>
 * All methods use {@link Location} for path arguments and throw {@link IOException}
 * on failure rather than returning a {@link org.apache.doris.backup.Status} object.
 * <p>
 * Implementations must be thread-safe. Instances may be shared and cached.
 *
 * @see LegacyFileSystemAdapter for adapting existing Status-based implementations
 * @see MemoryFileSystem for in-memory testing
 * @see LegacyFileSystemApi for the legacy Status-based interface (deprecated)
 */
public interface FileSystem extends Closeable {

    // ─────────────────────────── I/O FILE ACCESS ───────────────────────────

    /**
     * Creates a new {@link DorisInputFile} for reading from the given location.
     * Does not check existence; the file is opened lazily when
     * {@link DorisInputFile#newInput()} or {@link DorisInputFile#newStream()} is called.
     */
    DorisInputFile newInputFile(Location location);

    /**
     * Creates a new {@link DorisInputFile} with a known file length.
     * The {@code length} hint may be used to skip a remote {@code stat()} call.
     */
    DorisInputFile newInputFile(Location location, long length);

    /**
     * Creates a new {@link DorisOutputFile} for writing to the given location.
     */
    DorisOutputFile newOutputFile(Location location);

    // ─────────────────────────── METADATA OPERATIONS ───────────────────────────

    /**
     * Returns {@code true} if the location exists (file or directory).
     *
     * @throws IOException if the existence check fails for a reason other than "not found"
     */
    boolean exists(Location location) throws IOException;

    // ─────────────────────────── MUTATION OPERATIONS ───────────────────────────

    /**
     * Deletes the file at the given location.
     *
     * @throws IOException if deletion fails or the location is a directory
     */
    void deleteFile(Location location) throws IOException;

    /**
     * Deletes multiple files. Default implementation calls {@link #deleteFile} in a loop;
     * implementations may override for batch efficiency.
     */
    default void deleteFiles(Collection<Location> locations) throws IOException {
        for (Location location : locations) {
            deleteFile(location);
        }
    }

    /**
     * Atomically renames (moves) a file from {@code source} to {@code target}.
     *
     * @throws IOException if the rename fails
     */
    void renameFile(Location source, Location target) throws IOException;

    /**
     * Recursively deletes a directory and all its contents.
     *
     * @throws IOException if deletion fails
     */
    void deleteDirectory(Location location) throws IOException;

    /**
     * Creates a directory (and any missing parent directories).
     *
     * @throws IOException if creation fails
     */
    void createDirectory(Location location) throws IOException;

    /**
     * Renames (moves) a directory from {@code source} to {@code target}.
     *
     * @throws IOException if the rename fails
     */
    void renameDirectory(Location source, Location target) throws IOException;

    // ─────────────────────────── LISTING OPERATIONS ───────────────────────────

    /**
     * Returns a lazy iterator over files under {@code location}.
     * <p>
     * If {@code recursive} is {@code false}, only direct children are listed.
     * If {@code recursive} is {@code true}, the entire subtree is traversed.
     * The iterator must be closed after use (use try-with-resources).
     *
     * @throws IOException if the listing request fails
     */
    FileIterator listFiles(Location location, boolean recursive) throws IOException;

    /**
     * Returns the set of direct child directory locations under {@code location}.
     *
     * @throws IOException if the listing request fails
     */
    Set<Location> listDirectories(Location location) throws IOException;

    /**
     * Returns a lazy iterator of files matching the glob {@code pattern}.
     * Default implementation throws {@link UnsupportedOperationException}.
     * Override for native glob support (e.g., HDFS glob).
     */
    default FileIterator globFiles(Location pattern) throws IOException {
        throw new UnsupportedOperationException(
                "globFiles not supported by " + getClass().getSimpleName());
    }

    /**
     * Releases resources held by this filesystem (connections, thread pools, etc.).
     * Idempotent — safe to call multiple times.
     */
    @Override
    void close() throws IOException;
}
